From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Michael Goffioul <michael.goffioul@gmail.com>
Date: Wed, 28 Sep 2022 08:13:21 +0200
Subject: [PATCH 21/27] Add media extensions for FFMPEG codec2 service

A plain buffer is used to transfer extra information from extractor to
codecs. The buffer is mapped to a custom C2 paramkey.

Change-Id: Id9e36ae90b194b2973ed16af07a9e2b5230aaab7
Signed-off-by: SupeChicken666 <me@supechicken666.dev>
---
 media/codec2/core/include/C2Config.h                      | 8 ++++++++
 media/codec2/sfplugin/CCodecConfig.cpp                    | 5 +++++
 media/libstagefright/FFMPEGUtil.cpp                       | 1 +
 media/libstagefright/Utils.cpp                            | 1 +
 .../include/media/stagefright/MetaDataBase.h              | 1 +
 5 files changed, 16 insertions(+)

diff --git a/media/codec2/core/include/C2Config.h b/media/codec2/core/include/C2Config.h
index fbd1b36b8..eb28f2761 100644
--- a/media/codec2/core/include/C2Config.h
+++ b/media/codec2/core/include/C2Config.h
@@ -303,6 +303,9 @@ enum C2ParamIndexKind : C2Param::type_index_t {
     // input surface
     kParamIndexCaptureFrameRate, // input-surface, float
     kParamIndexStopTimeOffset, // input-surface, int64_t
+
+    // FFMPEG
+    kParamIndexRawCodecData,
 };
 
 }
@@ -2863,6 +2866,11 @@ typedef C2StreamParam<C2Info, C2SimpleValueStruct<int32_t>, kParamIndexAverageBl
         C2AndroidStreamAverageBlockQuantizationInfo;
 constexpr char C2_PARAMKEY_AVERAGE_QP[] = "coded.average-qp";
 
+// FFMEG
+
+typedef C2StreamParam<C2Info, C2BlobValue, kParamIndexRawCodecData> C2StreamRawCodecDataInfo;
+constexpr char C2_PARAMKEY_RAW_CODEC_DATA[] = "coded.raw-codec-data";
+
 /// @}
 
 #endif  // C2CONFIG_H_
diff --git a/media/codec2/sfplugin/CCodecConfig.cpp b/media/codec2/sfplugin/CCodecConfig.cpp
index 897a696f7..8aaa8a301 100644
--- a/media/codec2/sfplugin/CCodecConfig.cpp
+++ b/media/codec2/sfplugin/CCodecConfig.cpp
@@ -1052,6 +1052,11 @@ void CCodecConfig::initializeStandardParams() {
     KEY_AUDIO_SESSION_ID // we use "audio-hw-sync"
     KEY_OUTPUT_REORDER_DEPTH
     */
+
+    // FFMPEG
+
+    add(ConfigMapper("raw-codec-data", C2_PARAMKEY_RAW_CODEC_DATA, "value")
+        .limitTo(D::DECODER & D::CODED));
 }
 
 status_t CCodecConfig::initialize(
diff --git a/media/libstagefright/FFMPEGUtil.cpp b/media/libstagefright/FFMPEGUtil.cpp
index 76a765852..1958196b8 100644
--- a/media/libstagefright/FFMPEGUtil.cpp
+++ b/media/libstagefright/FFMPEGUtil.cpp
@@ -46,6 +46,7 @@ static const MetaKeyEntry MetaKeyTable[] {
    {kKeyCodecId              , "codec-id"               , INT32},
    {kKeyCodedSampleBits      , "coded-sample-bits"      , INT32},
    {kKeyFileFormat           , "file-format"            , INT32},
+   {kKeyRawCodecData         , "raw-codec-data"         , DATA},
    {kKeyRawCodecSpecificData , "raw-codec-specific-data", CSD},
    {kKeyPcmEncoding          , "pcm-encoding"           , INT32},
    {kKeyRVVersion            , "rv-version"             , INT32},
diff --git a/media/libstagefright/Utils.cpp b/media/libstagefright/Utils.cpp
index e553b9f62..8a336cfde 100644
--- a/media/libstagefright/Utils.cpp
+++ b/media/libstagefright/Utils.cpp
@@ -839,6 +839,7 @@ static std::vector<std::pair<const char *, uint32_t>> bufferMappings {
         { "slow-motion-markers", kKeySlowMotionMarkers },
         { "thumbnail-csd-av1c", kKeyThumbnailAV1C },
         // FFMEG
+        { "raw-codec-data", kKeyRawCodecData },
         { "raw-codec-specific-data", kKeyRawCodecSpecificData },
     }
 };
diff --git a/media/libstagefright/include/media/stagefright/MetaDataBase.h b/media/libstagefright/include/media/stagefright/MetaDataBase.h
index 872cdd9ab..913fb5fbe 100644
--- a/media/libstagefright/include/media/stagefright/MetaDataBase.h
+++ b/media/libstagefright/include/media/stagefright/MetaDataBase.h
@@ -153,6 +153,7 @@ enum {
 
     kKeyIsUnreadable      = 'unre',  // bool (int32_t)
 
+    kKeyRawCodecData         = 'rcdt',
     kKeyRawCodecSpecificData = 'rcsd',  // raw data - added to support mmParser
     kKeyDivXVersion       = 'DivX',  // int32_t
     kKeyDivXDrm           = 'QDrm',  // void *
